demos/selenium-test-app/lib.rs

```diff
@@ -0,0 +1,239 @@
+    });
+    let witness = asset_hashes.witness(asset_name.as_bytes());
+    let tree = ic_certified_map::labeled(b"http_assets", witness);
+    let mut serializer = serde_cbor::ser::Serializer::new(vec![]);
+    serializer.self_describe().unwrap();
+    tree.serialize(&mut serializer)
+        .unwrap_or_else(|e| api::trap(&format!("failed to serialize a hash tree: {}", e)));
+    (
+        "IC-Certificate".to_string(),
+        format!(
+            "certificate=:{}:, tree=:{}:",
+            base64::encode(&certificate),
+            base64::encode(&serializer.into_inner())
+        ),
+    )
+}
+
+#[init]
+#[post_upgrade]
+pub fn init_assets() {
+    ASSETS.with(|a| {
+        ASSET_HASHES.with(|ah| {
+            let mut assets = a.borrow_mut();
+            let mut asset_hashes = ah.borrow_mut();
+            for (path, content, content_type) in get_assets() {
+                asset_hashes.insert(path, sha2::Sha256::digest(content).into());
+                let mut headers = vec![];
+                headers.push((
+                    "Content-Type".to_string(),
+                    content_type.to_mime_type_string(),
+                ));
+                assets.insert(path, (headers, content.to_vec()));
+            }
+            update_root_hash(&asset_hashes);
+        });
+    });
+}
+
+fn update_root_hash(a: &RbTree<&'static str, [u8; 32]>) {
+    let prefixed_root_hash = labeled_hash(b"http_assets", &a.root_hash());
+    api::set_certified_data(&prefixed_root_hash[..]);
+}
+
+lazy_static! {
+    // The full content of the index.html, after the canister ID (and script tag) have been
+    // injected
+    static ref INDEX_HTML_STR: String = {
+        let canister_id = api::id();
+        let index_html = include_str!("dist/index.html");
+        let index_html = index_html.replace(
+            r#"<script defer="defer" src="bundle.js"></script>"#,
+            &format!(r#"<script id="setupJs">var canisterId = '{canister_id}';</script><script defer="defer" src="bundle.js"></script>"#).to_string()
+        );
+        index_html
+    };
+}
+
+// Get all the assets. Duplicated assets like index.html are shared and generally all assets are
+// prepared only once (like injecting the canister ID).
+fn get_assets() -> [(&'static str, &'static [u8], ContentType); 5] {
+    let index_html: &[u8] = INDEX_HTML_STR.as_bytes();
+    [
+        ("/", index_html, ContentType::HTML),
+        ("/index.html", index_html, ContentType::HTML),
+        (
+            "/bundle.js",
+            include_bytes!("dist/bundle.js"),
+            ContentType::JS,
+        ),
+        (
+            "/.well-known/ii-alternative-origins",
+            b"{\"alternativeOrigins\":[]}",
+            ContentType::JSON,
+        ),
+        (
+            "/.well-known/evil-alternative-origins",
+            b"{\"alternativeOrigins\":[\"https://evil.com\"]}",
+            ContentType::JSON,
+        ),
```

Comment:
Can you explain these two values? Why do we return an empty list of origins, and when would `evil-...` be queried?

Reply:
The empty list is just the default. The value of this asset can be updated by calling `update_alternative_origins`.
`/.well-known/evil-alternative-origins` is just the default value for the `redirect` mode explained in the previous comment.

Reply:
I see. Say someone were to discover the codebase for the first time, how easy do you reckon it would be to figure all of this out? Should we add some notes about this somewhere, for instance here where you define the assets?

Reply:
Yes, let me do a clean-up pass through the test app. I have created L2-885 to track this.